<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">

    <xacro:include filename="consts.xacro" /> 

    <xacro:macro name="left_ext_first_joint_limits">
        <limit lower="${left_first_link_low}" upper="${left_first_link_up}" velocity="${velocity}" effort="${effort}"/> <!-- effort = 7 -->
    </xacro:macro>
    <xacro:macro name="right_ext_first_joint_limits">
        <limit lower="${right_first_link_low}" upper="${right_first_link_up}" velocity="${velocity}" effort="${effort}"/> <!-- effort = 7 -->
    </xacro:macro>
    <xacro:macro name="left_first_sec_joint_limits">
        <limit lower="${left_sec_link_low}" upper="${left_sec_link_up}" velocity="${velocity}" effort="${effort}"/>
    </xacro:macro>
    <xacro:macro name="right_first_sec_joint_limits">
        <limit lower="${right_sec_link_low}" upper="${right_sec_link_up}" velocity="${velocity}" effort="${effort}"/>
    </xacro:macro>


<!-- !!!!!!!!!!!!!!!!!!       fixed_joint_ext         !!!!!!!!!!! -->


    <joint name="base_footprint_base" type="fixed">
        <origin xyz="0 0 0" rpy = "0 0 3.925"/>
        <parent link="base_footprint"/>
        <child link="base_link"/>
    </joint>

    <joint name="base_frontleftext" type="fixed">
        <origin xyz="-0.085989 0.009337 0.2" rpy="0 0 0"/>
        <parent link="base_link"/>
        <child link="front_left_ext"/>
    </joint>

    <joint name="base_backleftext" type="fixed">
        <origin xyz="0.001768 -0.07869 0.2" rpy="0 0 0"/>
        <parent link="base_link"/>
        <child link="back_left_ext"/>
    </joint>

    <joint name="base_frontrightext" type="fixed">
        <origin xyz="-0.010606 0.091144 0.2" rpy="0 0 0"/>
        <parent link="base_link"/>
        <child link="front_right_ext"/>
    </joint>

    <joint name="base_backrightext" type="fixed">
        <origin xyz="0.077977 0.002501 0.2" rpy="0 0 0"/>
        <parent link="base_link"/>
        <child link="back_right_ext"/>
    </joint>


<!-- !!!!!!!!!!!!!!!!!!       ext_first_joints         !!!!!!!!!!! -->



    <joint name="ext_frontleftfirst" type="revolute">
        <origin xyz="-0.021207 -0.014909 -0.0735" rpy="0 0 0"/>
        <parent link="front_left_ext"/>
        <child link="front_left_first"/>
        <axis xyz="0.0 -0.0 1.0"/>
        <xacro:left_ext_first_joint_limits></xacro:left_ext_first_joint_limits>
        <xacro:if value="${use_dynamics}">
            <xacro:ext_first_joint_dynamics></xacro:ext_first_joint_dynamics>
        </xacro:if>    
    </joint>

    <joint name="ext_backleftfirst" type="revolute">
        <origin xyz="-0.021207 -0.01491 -0.0735" rpy="0 0 0"/>
        <parent link="back_left_ext"/>
        <child link="back_left_first"/>
        <axis xyz="0.0 -0.0 1.0"/>
        <xacro:left_ext_first_joint_limits></xacro:left_ext_first_joint_limits>
        <xacro:if value="${use_dynamics}">
            <xacro:ext_first_joint_dynamics></xacro:ext_first_joint_dynamics>
        </xacro:if>    
    </joint>

    <joint name="ext_backrightfirst" type="revolute">
        <origin xyz="0.015989 0.016617 -0.0735" rpy="0 0 0"/>
        <parent link="back_right_ext"/>
        <child link="back_right_first"/>
        <axis xyz="0.0 -0.0 1.0"/>
        <xacro:right_ext_first_joint_limits></xacro:right_ext_first_joint_limits>
        <xacro:if value="${use_dynamics}">
            <xacro:ext_first_joint_dynamics></xacro:ext_first_joint_dynamics>
        </xacro:if>    
    </joint>

    <joint name="ext_frontrightfirst" type="revolute">
        <origin xyz="0.015989 0.016617 -0.0735" rpy="0 0 0"/>
        <parent link="front_right_ext"/>
        <child link="front_right_first"/>
        <axis xyz="0.0 -0.0 1.0"/>
        <xacro:right_ext_first_joint_limits></xacro:right_ext_first_joint_limits>
        <xacro:if value="${use_dynamics}">
            <xacro:ext_first_joint_dynamics></xacro:ext_first_joint_dynamics>
        </xacro:if>     
    </joint>



<!-- !!!!!!!!!!!!!!!!!!      first_SEC_joints         !!!!!!!!!!! -->



    <joint name="first_frontleftsec" type="revolute">
        <origin xyz="-0.019366 -0.06745 0.017042" rpy="0 0 0"/>
        <parent link="front_left_first"/>
        <child link="front_left_sec"/>
        <axis xyz="0.707107 -0.707107 0.0"/>
        <xacro:left_first_sec_joint_limits></xacro:left_first_sec_joint_limits>
        <xacro:if value="${use_dynamics}">
            <xacro:first_sec_joint_dynamics></xacro:first_sec_joint_dynamics>
        </xacro:if>     
    </joint>

    <joint name="first_backleftsec" type="revolute">
        <origin xyz="-0.019366 -0.067449 0.017042" rpy="0 0 0"/>
        <parent link="back_left_first"/>
        <child link="back_left_sec"/>
        <axis xyz="0.707107 -0.707107 0.0"/>
        <xacro:left_first_sec_joint_limits></xacro:left_first_sec_joint_limits>
        <xacro:if value="${use_dynamics}">
            <xacro:first_sec_joint_dynamics></xacro:first_sec_joint_dynamics>
        </xacro:if>      
    </joint>

    <joint name="first_frontrightsec" type="revolute">
        <origin xyz="0.067449 0.019366 0.017042" rpy="0 0 0"/>
        <parent link="front_right_first"/>
        <child link="front_right_sec"/>
        <axis xyz="0.707107 -0.707107 -0.0"/>
        <xacro:right_first_sec_joint_limits></xacro:right_first_sec_joint_limits>
        <xacro:if value="${use_dynamics}">
            <xacro:first_sec_joint_dynamics></xacro:first_sec_joint_dynamics>
        </xacro:if>      
    </joint>

    <joint name="first_backrightsec" type="revolute">
        <origin xyz="0.067449 0.019366 0.017042" rpy="0 0 0"/>
        <parent link="back_right_first"/>
        <child link="back_right_sec"/>
        <axis xyz="0.707107 -0.707107 0.0"/>
        <xacro:right_first_sec_joint_limits></xacro:right_first_sec_joint_limits>
        <xacro:if value="${use_dynamics}">
            <xacro:first_sec_joint_dynamics></xacro:first_sec_joint_dynamics>
        </xacro:if>     
    </joint>



<!-- !!!!!!!!!!!!!!!!!!      base_wing_joints         !!!!!!!!!!! -->



    <joint name="base_frontleftwing" type="revolute">
        <origin xyz="-0.1425 4e-05 0.2645" rpy="0 0 0"/>
        <parent link="base_link"/>
        <child link="front_left_wing"/>
        <axis xyz="0.0 -0.0 1.0"/>
        <limit upper="1.570796" lower="-1.570796" effort="100" velocity="100"/>
    </joint>

    <joint name="base_frontrightwing" type="revolute">
        <origin xyz="0.0 0.142742 0.265" rpy="0 0 0"/>
        <parent link="base_link"/>
        <child link="front_right_wing"/>
        <axis xyz="0.0 -0.0 1.0"/>
        <limit upper="1.570796" lower="-1.570796" effort="100" velocity="100"/>
    </joint>

    <joint name="base_backleftwing" type="revolute">
        <origin xyz="0.0 -0.142742 0.265" rpy="0 0 0"/>
        <parent link="base_link"/>
        <child link="back_left_wing"/>
        <axis xyz="0.0 -0.0 1.0"/>
        <limit upper="1.570796" lower="-1.570796" effort="100" velocity="100"/>
    </joint>

    <joint name="base_backrightwing" type="revolute">
        <origin xyz="0.1425 4e-05 0.264887" rpy="0 0 0"/>
        <parent link="base_link"/>
        <child link="back_right_wing"/>
        <axis xyz="0.0 -0.0 1.0"/>
        <limit upper="1.570796" lower="-1.570796" effort="100" velocity="100"/>
    </joint>



<!-- !!!!!!!!!!!!!!!!!!      wing_propeller_joints         !!!!!!!!!!! -->



    <joint name="wing_backleftprop" type="continuous">
        <origin xyz="-1e-05 -0.231282 -0.01" rpy="0 0 0"/>
        <parent link="back_left_wing"/>
        <child link="back_left_prop"/>
        <axis xyz="0.0 -0.0 1.0"/>
    </joint>

    <joint name="wing_backrightprop" type="continuous">
        <origin xyz="0.231282 -1e-05 -0.01" rpy="0 0 0"/>
        <parent link="back_right_wing"/>
        <child link="back_right_prop"/>
        <axis xyz="0.0 -0.0 1.0"/>
    </joint>

    <joint name="wing_frontrightprop" type="continuous">
        <origin xyz="1e-05 0.231282 -0.01" rpy="0 0 0"/>
        <parent link="front_right_wing"/>
        <child link="front_right_prop"/>
        <axis xyz="0.0 -0.0 1.0"/>
    </joint>

    <joint name="wing_frontleftprop" type="continuous">
        <origin xyz="-0.231282 1e-05 -0.01" rpy="0 0 0"/>
        <parent link="front_left_wing"/>
        <child link="front_left_prop"/>
        <axis xyz="0.0 -0.0 1.0"/>
    </joint>



<!-- !!!!!!!!!!!!!!!!!!      SEC_foot_joints         !!!!!!!!!!! -->



    <joint name="FR_foot_fixed" type="fixed">
        <origin rpy="0 0 0" xyz="-0.0165 0.0165 -0.128"/>
        <parent link="front_right_sec"/>
        <child link="FR_foot"/>
    </joint>
     <!-- *********  FL_FOOT  ********* -->
    <joint name="FL_foot_fixed" type="fixed">
        <origin rpy="0 0 0" xyz="-0.0165 0.0165 -0.128"/>
        <parent link="front_left_sec"/>
        <child link="FL_foot"/>
    </joint>
     <!-- *********  BR_FOOT  ********* -->
    <joint name="BR_foot_fixed" type="fixed">
        <origin rpy="0 0 0" xyz="-0.0165 0.0165 -0.128"/>
        <parent link="back_right_sec"/>
        <child link="BR_foot"/>
    </joint>
    <!-- *********  BL_FOOT  ********* -->
    <joint name="BL_foot_fixed" type="fixed">
        <origin rpy="0 0 0" xyz="-0.0165 0.0165 -0.128"/>
        <parent link="back_left_sec"/>
        <child link="BL_foot"/>
    </joint>

</robot>